rust/agama-server/src/agama-web-server.rs

```diff
@@ -32,23 +60,213 @@ struct Cli {
+                if let Err(err) = SslStream::accept(Pin::new(&mut tls_stream)).await {
+                    tracing::error!("Error during TSL handshake from {}: {}", addr, err);
+                }
+
+                let stream = TokioIo::new(tls_stream);
+                let hyper_service =
+                    hyper::service::service_fn(move |request: Request<Incoming>| {
+                        tower_service.clone().call(request)
+                    });
+
+                let ret = hyper_util::server::conn::auto::Builder::new(TokioExecutor::new())
+                    .serve_connection_with_upgrades(stream, hyper_service)
+                    .await;
+
+                if let Err(err) = ret {
+                    tracing::error!("Error serving connection from {}: {}", addr, err);
+                }
+            } else {
+                // handle HTTP connection
+                let stream = TokioIo::new(tcp_stream);
+                let hyper_service =
+                    hyper::service::service_fn(move |request: Request<Incoming>| {
+                        // check if it is local connection or external
+                        // the to_canonical() converts IPv4-mapped IPv6 addresses
+                        // to plain IPv4, then is_loopback() works correctly for the IPv4 connections
+                        if addr.ip().to_canonical().is_loopback() {
+                            // accept plain HTTP on the local connection
+                            tower_service.clone().call(request)
+                        } else {
+                            // redirect external connections to HTTPS
+                            redirector_service.clone().call(request)
+                        }
+                    });
+
+                let ret = hyper_util::server::conn::auto::Builder::new(TokioExecutor::new())
+                    .serve_connection_with_upgrades(stream, hyper_service)
+                    .await;
+
+                if let Err(err) = ret {
+                    tracing::error!("Error serving connection from {}: {}", addr, err);
+                }
+            }
+        });
+    }
+}
+
 /// Start serving the API.
-async fn serve_command(address: &str) -> anyhow::Result<()> {
+async fn serve_command(
+    address: &str,
+    address2: &str,
+    cert: &String,
+    key: &String,
+) -> anyhow::Result<()> {
     let journald = tracing_journald::layer().expect("could not connect to journald");
     tracing_subscriber::registry().with(journald).init();
 
-    let listener = tokio::net::TcpListener::bind(address)
-        .await
-        .unwrap_or_else(|_| panic!("could not listen on {}", address));
-
     let (tx, _) = channel(16);
     run_monitor(tx.clone()).await?;
 
     let config = web::ServiceConfig::load().unwrap();
     let service = web::service(config, tx);
-    axum::serve(listener, service)
-        .await
-        .expect("could not mount app on listener");
+    let ssl_acceptor = create_ssl_acceptor(cert, key);
```

Comment:
Setting up the SSL Acceptor looks like a job for the `start_server` function. If you had a `Certificate` struct it would feel kind of natural to pass it as an argument to the `start_server` function.